test/e2e/virtual-routes.test.js

```diff
@@ -0,0 +1,284 @@
+const docsifyInit = require('../helpers/docsify-init');
+const { navigateToRoute } = require('../helpers/navigate');
+const { test, expect } = require('./fixtures/docsify-init-fixture');
+
```

Comment:
Can we add a basic test using `routerMode: 'history'` to verify that routes and the `navigateToRoute()` helper works as expected?

Reply:
ok so seems like `docsifyInit` cannot create test pages that use `routerMode: history`:
![image](https://user-images.githubusercontent.com/6681893/169371509-a7ae3243-4607-45c3-b433-16c07d7e902f.png)
![image](https://user-images.githubusercontent.com/6681893/169371604-27434437-9edf-41a0-81de-76ea9895033e.png)

I can fix it, but IMO it seems a little out of scope for this PR. How do we feel about moving the `navigateToPage` helper into the `virtual-routes.test.js` file and explicitly supporting only `hash` mode?

Reply:
Agreed on fixing `docsifyInit()` being out of scope.

Also agree on the simple solution: move the `navigateToRoute()` helper into the `virtual-routes.test.js`.

However, if you are motivated to do so you could verify that `navigateToRoute()` works as expected on docsify sites using `routeMode: 'history'` when `docsifyInit()` isn't used. Assuming this work, this would allow the helper to remain in the `/test/helpers` directory so it can be used today when testing against the `/docs/` site or in the future if/when we fix `docsifyInit()`.

I gave this a shot myself by setting `routerMode: 'history'` in `/docs/index.html` and using the following test (WIP):

```js
test('routerMode:history', async ({ page }) => {
  // Docs site
  page.goto('http://localhost:3001/docs/');

  // Wait for docsify to initialize
  await page.waitForSelector('#main > *');
  await page.waitForLoadState('networkidle');

  // Verify routerMode
  const $docsify = await page.evaluate(() => window.$docsify);
  expect($docsify).toHaveProperty('routerMode', 'history');

  // URL Before
  console.log(page.url());

  // Navigate
  navigateToRoute(page, '/quickstart');

  // URL Before
  console.log(page.url());
});
```

The above test did not work for me for two reasons:

1. First, the `route` wasn't being passed to `page.evaluate` when the `routeMode` was set to `history`. Fixed that (below).
   ```js
   // Before
   if (routerMode === 'history') {
     await page.evaluate(r => window.history.pushState({ key: r }, '', r));
   }

   // After
   if (routerMode === 'history') {
     await page.evaluate(
       r => window.history.pushState({ key: r }, '', r),
       route
     );
   }
   ```
1. After calling `history.pushState`, the I received the following Playwright error:
   ```
   page.evaluate: Target closed
   ```

I started down the rabbit hole and began playing `page.waitForNavigation()` and researching how to use `history.pushState` without closing the Playwright `page` object, then I realized I was going down the rabbit hole and stopped myself. :smile:

Anyway, no need to go down this rabbit hole yourself if you aren't interested. Just sharing my results after tinkering for a bit in case you wanted to continue the effort.

Reply:
Interesting. Perhaps pushing history state is considered by the driver as navigation, therefore destroying the context in which the `evaluate` method was called. I'll try to take a look there; in the meantime, I moved the function into the test file :)